File: tasks/getCapabilities.py
#!/usr/bin/env python

from optparse import OptionParser
from datetime import datetime
from collections import OrderedDict
import os
import sys
import json
import urllib2 as urllib
import xmltodict
import traceback

prog = os.path.basename(__file__)
base_dir = os.path.join(os.path.dirname(__file__), "..")
version = "1.0.0"
help_description = """\
Fetches options from remote locations
"""

parser = OptionParser(usage="Usage: %s <config> <output_dir>" % prog,
                      version="%s version %s" % (prog, version),
                      epilog=help_description)

(options, args) = parser.parse_args()
if len(args) != 2:
    parser.error("Invalid number of arguments")

config_file = args[0]
output_dir = args[1]
colormaps = {}
colormaps_dir = os.path.join(output_dir, "colormaps")
remote_count = 0
error_count = 0
warning_count = 0

if not os.path.exists(output_dir):
    os.makedirs(output_dir)

with open(config_file) as fp:
    config = json.load(fp)

def process_layer(layer):
    ident = layer["ows:Identifier"]
    if "ows:Metadata" in layer:
        if ident in config.get("skipPalettes", []):
            sys.stderr.write("%s:    WARN: Skipping palette for %s\n" %
                             prog, ident)
            global warning_count
            warning_count += 1
        else:
            for item in layer["ows:Metadata"]:
                schema_version = item["@xlink:role"]
                if schema_version == "http://earthdata.nasa.gov/gibs/metadata-type/colormap/1.3":
                    colormap_link = item["@xlink:href"]
                    #colormap_link = layer["ows:Metadata"]["@xlink:href"]
                    colormap_file = os.path.basename(colormap_link)
                    colormap_id = os.path.splitext(colormap_file)[0]
                    colormaps[colormap_id] = colormap_link

def process_remote(entry):
    url = entry["from"]
    print "%s: %s" % (prog, url)
    response = urllib.urlopen(url)
    contents = response.read()
    output_file = os.path.join(output_dir, entry["to"])

    # Write GetCapabilities responses to XML files
    with open(output_file, "w") as fp:
        fp.write(contents)
    gc = xmltodict.parse(contents)

    # Find all colormaps in GetCapabilities responses and store them in memory
    try:
        if(type(gc["Capabilities"]["Contents"]["Layer"]) is OrderedDict):
            process_layer(gc["Capabilities"]["Contents"]["Layer"])
        else:
            for layer in gc["Capabilities"]["Contents"]["Layer"]:
                process_layer(layer)

    except:
        print(ident)
        print(str(traceback.format_exc()))

# Fetch every colormap from the API and write response to file system
def process_colormaps():
    print "%s: Fetching %d colormaps" % (prog, len(colormaps))
    sys.stdout.flush()
    if not os.path.exists(colormaps_dir):
        os.makedirs(colormaps_dir)
    for link in colormaps.values():
        try:
            response = urllib.urlopen(link)
            contents = response.read()
            output_file = os.path.join(colormaps_dir, os.path.basename(link))
            with open(output_file, "w") as fp:
                fp.write(contents)
        except Exception as e:
            sys.stderr.write("%s:   WARN: Unable to fetch %s: %s" %
                (prog, link, str(e)))
            global warning_count
            warning_count += 1
    print "%s: Fetching %d colormaps" % (prog, len(colormaps))

tolerant = config.get("tolerant", False)
if "wv-options-fetch" in config:
    for entry in config["wv-options-fetch"]:
        try:
            remote_count += 1
            process_remote(entry)
        except Exception as e:
            if tolerant:
                warning_count += 1
                sys.stderr.write("%s:   WARN: %s\n" % (prog, str(e)))
            else:
                error_count += 1
                sys.stderr.write("%s: ERROR: %s\n" % (prog, str(e)))
    if colormaps:
        process_colormaps()

print "%s: %d error(s), %d remote(s)" % (prog, error_count, remote_count)

if error_count > 0:
    sys.exit(1)

File: tasks/node/getCapabilities.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const yargs = require('yargs');
const console = require('console');
const request = require('axios').default;
const convert = require('xml-js');
const { promisify } = require('util');
const stream = require('stream');

const finished = promisify(stream.finished);
const prog = path.basename(__filename);
// const baseDir = path.join(__dirname, '..');

const options = yargs
  .usage('Usage: $0 [options]')
  .option('config', {
    demandOption: true,
    alias: 'c',
    type: 'string',
    description: 'config file',
  })
  .option('getcapabilities', {
    demandOption: true,
    alias: 'g',
    type: 'string',
    description: 'getcapabilities file',
  })
  .epilog('Pulls GetCapabilities XML and linked metadata from configured locations');

// import config file from ./config/default/release/config.json

const { argv } = options;
if (!argv.config && !argv.getcapabilities) {
  console.error('Invalid number of arguments');
  process.exit(1);
}
const configFile = argv.config;
const outputDir = argv.getcapabilities;
const colormaps = {};
const vectorstyles = {};
const vectordata = {};
const colormapsDir = path.join(outputDir, 'colormaps');
const vectorstylesDir = path.join(outputDir, 'vectorstyles');
const vectordataDir = path.join(outputDir, 'vectordata');

const configData = fs.readFileSync(configFile);
const config = JSON.parse(configData);

async function main () {
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir);
  }

  await GetCapabilities();

  // process.exit();
}

async function GetCapabilities() {
  // Download each GC xml using the "from" attribute and put it in the "to" location
  if (Object.prototype.hasOwnProperty.call(config, 'wv-options-fetch')) {
    const fetchValues = config['wv-options-fetch'];
    Object.entries(fetchValues).forEach(async ([key, value]) => {
      const inputFile = value.from;
      const outputFile = `${outputDir}/${value.to}`;

      await fetchConfigs(inputFile, outputFile);
      await processGetCapabilities(outputFile);

      if (colormaps) {
        await gatherProcess(colormaps, 'colormaps', colormapsDir, '.xml');
      }
      if (vectorstyles) {
        await gatherProcess(vectorstyles, 'vectorstyles', vectorstylesDir, '.json');
      }
      if (vectordata) {
        await gatherProcess(vectordata, 'vectordatas', vectordataDir, '.json');
      }
    });
  }
}

async function fetchConfigs(inputFile, outputFile) {
  const writer = await fs.createWriteStream(outputFile);
  return request({
    method: 'get',
    url: inputFile,
    responseType: 'stream',
  }).then(async (response) => {
    await response.data.pipe(writer);
    return finished(writer);
  });
}

async function handleException(error, link) {
  console.warn(`\n ${prog} WARN: Unable to fetch ${link} ${error}`);
}

async function processVectorData(layer) {
  if (layer['ows:Metadata']) {
    Object.values(layer['ows:Metadata']).forEach((item) => {
      const schemaVersion = item._attributes['xlink:role'];
      if (schemaVersion === 'http://earthdata.nasa.gov/gibs/metadata-type/layer/1.0') {
        const vectorDataLink = item._attributes['xlink:href'];
        const vectorDataFile = path.basename(vectorDataLink);
        const vectorDataId = path.parse(vectorDataFile).name;
        vectordata[vectorDataId] = vectorDataLink;
      }
    });
  }
}

async function processLayer(layer) {
  const ident = layer['ows:Identifier']._text;
  if (layer['ows:Metadata']) {
    if (config.skipPalettes) {
      console.warn(`${prog}: WARN: Skipping palette for ${ident} \n`);
    } else {
      Object.values(layer['ows:Metadata']).forEach((item) => {
        const schemaVersion = item._attributes['xlink:role'];
        if (schemaVersion === 'http://earthdata.nasa.gov/gibs/metadata-type/colormap/1.3') {
          const colormapLink = item._attributes['xlink:href'];
          const colormapFile = path.basename(colormapLink);
          const colormapId = path.parse(colormapFile).name;
          colormaps[colormapId] = colormapLink;
        } else if (schemaVersion === 'http://earthdata.nasa.gov/gibs/metadata-type/mapbox-gl-style/1.0') {
          const vectorStyleLink = item._attributes['xlink:href'];
          const vectorStyleFile = path.basename(vectorStyleLink);
          const vectorStyleId = path.parse(vectorStyleFile).name;
          vectorstyles[vectorStyleId] = vectorStyleLink;
        }
      });
    }
  }
}

async function processGetCapabilities(outputFile) {
  const gcXML = fs.readFileSync(outputFile, { encoding: 'utf-8' });
  let gc = convert.xml2json(gcXML, { compact: true, spaces: 2 });
  gc = JSON.parse(gc);

  try {
    if (!gc.Capabilities || !gc.Capabilities.Contents) {
      console.error(`error: ${outputFile}: no layers`);
      process.exit(1);
    }

    const layers = gc.Capabilities.Contents.Layer;

    Object.values(layers).forEach((layer) => {
      processLayer(layer);
      processVectorData(layer);
    });
  } catch (error) {
    console.error(`ERROR: ${outputFile}: ${error}`);
    process.exit(1);
  }
}

async function processMetadata(link, dir, ext) {
  try {
    const outputFile = path.join(dir, path.basename(link));
    if (link.endsWith(ext)) {
      await fetchConfigs(link, outputFile);
    }
  } catch (error) {
    handleException(error, link);
  }
}

async function gatherProcess(type, typeStr, dir, ext) {
  console.warn(`${prog}: Fetching ${Object.keys(type).length} ${typeStr}`);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir);
  }
  Object.values(type).forEach(async (link) => {
    await processMetadata(link, dir, ext);
  });
}

main().catch((err) => {
  console.error(err.stack);
});
